### PR TITLE
MULE-8865: Migrate Oauth module to extensions API

### PR DESCRIPTION
**Pull request by elrodro83**

### CODE REVIEW

**Comment by elrodro83** [comment]
jenkins validate

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/tokenmanager/TokenManagerConfig.java [hunk 1]
```diff
@@ -23,15 +30,34 @@
  *
  * It can be referenced to access the state inside a flow for custom processing of oauth dance content.
  */
+@Alias("token-manager-config")
+@XmlHints(allowTopLevelDefinition = true)
 public class TokenManagerConfig implements Initialisable, MuleContextAware {
```
Comment (marianogonzalez) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/tokenmanager/TokenManagerConfig.java]: This class should die and its content merge in OAuthExtension.java

File: distributions/standalone/assembly.xml [hunk 2]
```diff
@@ -150,7 +150,6 @@
                 <include>org.mule.modules:mule-module-jbossts</include>
                 <include>org.mule.modules:mule-module-json</include>
                 <include>org.mule.modules:mule-module-management</include>
-                <include>org.mule.modules:mule-module-oauth</include>
```
Comment (elrodro83) [inline distributions/standalone/assembly.xml]: this removes the old module. The new extension is put in the distro the same way other extensions are

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/OAuthExtension.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * An extension to hook oauth2 to http extension connectors.
+ *
+ * @since 4.0
+ */
+@Extension(name = "OAuth")
+@Import(type = HttpAuthentication.class, from = "HTTP")
+@Operations(OAuthOperations.class)
+@SubTypeMapping(baseType = HttpAuthentication.class,
+    subTypes = {DefaultAuthorizationCodeGrantType.class, ClientCredentialsGrantType.class})
+@Xml(namespace = "oauth")
```
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/OAuthExtension.java]: Removing this annotation causes the namespace prefix to be "o-auth"

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/DefaultAuthorizationCodeGrantType.java [hunk 4]
```diff
@@ -170,25 +187,17 @@ public TlsContextFactory getTlsContext() {
...
-        this.tokenManagerConfig = TokenManagerConfig.createDefault(muleContext);
-        this.tokenManagerConfig.initialise();
-      }
-      if (localAuthorizationUrlResourceOwnerIdEvaluator == null) {
-        localAuthorizationUrlResourceOwnerIdEvaluator = new AttributeEvaluator(null);
-      }
-      localAuthorizationUrlResourceOwnerIdEvaluator.initialize(muleContext.getExpressionManager());
-      if (resourceOwnerIdEvaluator == null) {
-        resourceOwnerIdEvaluator = new AttributeEvaluator(ResourceOwnerOAuthContext.DEFAULT_RESOURCE_OWNER_ID);
+      this.serversManager = muleContext.getRegistry().lookupObject(OAuthCallbackServersManager.class);
```
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/DefaultAuthorizationCodeGrantType.java]: doesn't work, created a JIRA and will add TODOs

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/DefaultAuthorizationCodeGrantType.java [hunk 5]
```diff
@@ -247,53 +258,34 @@ private void buildHttpServer() throws InitialisationException {
...
-                                                           getRefreshTokenWhen(), value));
-      }
-      Boolean shouldRetryRequest = (Boolean) value;
-      if (shouldRetryRequest) {
-        try {
-          refreshToken(firstAttemptResponseEvent, resourceOwnerIdEvaluator.resolveStringValue(firstAttemptResponseEvent));
-        } catch (MuleException e) {
-          throw new MuleRuntimeException(e);
-        }
+    Boolean shouldRetryRequest = resolveExpression(tokenRequestHandler.getRefreshTokenWhen(), firstAttemptResponseEvent);
```
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/DefaultAuthorizationCodeGrantType.java]: That is already being handled in AbstractAuthorizationCodeTokenRequestHandler#refreshToken

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/state/ConfigOAuthContext.java [hunk 6]
```diff
@@ -6,8 +6,8 @@
  */
 package org.mule.extension.oauth2.internal.authorizationcode.state;
 
-import org.mule.runtime.core.api.store.ListableObjectStore;
 import org.mule.runtime.core.api.lock.LockFactory;
+import org.mule.runtime.core.api.store.ListableObjectStore;
```
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/state/ConfigOAuthContext.java]: `¯\_(ツ)_/¯`

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/tokenmanager/TokenManagerConfig.java (hunk 1 above)
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/tokenmanager/TokenManagerConfig.java]: I tried doing that, but i stumbled upon an issue similar to the one where the `@Inject` for mule context doesn't work.
In this case, the `@UseConfig` annotation is ignored, so i cannot get the config from the HttpAuthentication implementation. Will create a JIRA to do this after this other issue is fixed.

File: extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/OAuthOperations.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ */
+public class OAuthOperations {
+
+  /**
+   * Clears the oauth context for a token manager and a resource owner id.
+   * 
+   * @param tokenManager The token manager which holds the credentials to invalidate.
+   * @param resourceOwnerId The resource owner id to invalidate. This attribute is only allowed for authorization code grant type.
+   */
+  public void invalidateOauthContext(TokenManagerConfig tokenManager,
```
Comment (elrodro83) [inline extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/OAuthOperations.java]: See https://github.com/mulesoft/mule/pull/3941#discussion-diff-95906036R35

**Pull request merged**